spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java

```diff
@@ -495,16 +495,24 @@ public void visitTupleExpression(TupleExpression expr) {
 
   private Expression record(Expression expr) {
     // replace expr with $spock_valueRecorder.record($spock_valueRecorder.startRecordingValue(recordCount++), <expr>)
-    return AstUtil.createDirectMethodCall(
-        new VariableExpression(SpockNames.VALUE_RECORDER),
-        resources.getAstNodeCache().ValueRecorder_Record,
-        new ArgumentListExpression(
-            AstUtil.createDirectMethodCall(
-                new VariableExpression(SpockNames.VALUE_RECORDER),
-                resources.getAstNodeCache().ValueRecorder_StartRecordingValue,
-                new ArgumentListExpression(new ConstantExpression(recordCount++))
-            ),
-            expr));
+    MethodCallExpression result = AstUtil.createDirectMethodCall(
+      new VariableExpression(SpockNames.VALUE_RECORDER),
+      resources.getAstNodeCache().ValueRecorder_Record,
+      new ArgumentListExpression(
+        AstUtil.createDirectMethodCall(
+          new VariableExpression(SpockNames.VALUE_RECORDER),
+          resources.getAstNodeCache().ValueRecorder_StartRecordingValue,
+          new ArgumentListExpression(new ConstantExpression(recordCount++))
+        ),
+        expr));
+
+    // maintain type casts to preserve method disambiguation
+    // for example when giving a casted null as argument
+    if (expr instanceof CastExpression) {
```

Comment:
This code deals with casts in conditions, but in `DisambiguationInConditions` I only see casts in the `when` and not the `then`/`expect` blocks

Reply:
Wtf, you are right.
Usually I verify that my tests fail without my changes.
I don't know what I f**** up there, I need to rewrite those tests, thanks for catching.

Reply:
Ah, I had them first similar to how I reported it with `System.out.println` and so on, had them failing, fixed them and then refactored (and thereby broke) the tests to properly verify the result.
The tests are fixed now.